File: DynamicProgramming/JumpGame/jumpGame.java
class JumpGame {
  public boolean canJump(int[] nums) {
    int n = nums.length;
    int farthest = 0;

    for (int i = 0; i < n - 1; i++) {
      if (nums[i] + i > farthest) {
        farthest = nums[i] + i;
      }
      if (farthest <= i)
        return false;
    }

    return farthest >= n - 1;
  }

  public int jump(int[] nums) {
    int n = nums.length;
    int end = 0;
    int farthest = 0;
    int jumps = 0;

    for (int i = 0; i < n - 1; i++) {
      farthest = (nums[i] + i) > farthest ? (nums[i] + i) : farthest;

      if (end == i) {
        jumps++;
        end = farthest;
        // 若题目不保证一定能到终点, 则需要判断终止条件 (最远的距离所在的可跳步数是 0)
        // if (farthest == i) {
        //   return -1;
        // }
      }
    }

    return jumps;
  }

  public static void main(String[] args) {
    JumpGame jg = new JumpGame();
    int[] nums = {2, 3, 1, 1, 4};
    int[] _nums = {3, 2, 1, 0, 4};
    boolean res = jg.canJump(nums);
    boolean _res = jg.canJump(_nums);
    System.out.println(res);
    System.out.println(_res);
    System.out.println(jg.jump(nums));
  }
}

File: DynamicProgramming/JumpGame/jumpGame.py
class JumpGame(object):
    def canJump(self, nums) -> bool:
        n = len(nums)
        farthest = 0

        for i in range(n - 1):
            farthest = farthest if farthest > (i + nums[i]) else (i + nums[i])

            if (farthest <= i):
                return False
        
        return farthest >= (n - 1)

    def jump(self, nums) -> int:
        n = len(nums)
        farthest = 0
        end = 0
        jumps = 0

        for i in range(n - 1):
            farthest = farthest if farthest > (i + nums[i]) else (i + nums[i])

            if (i == end):
                jumps += 1
                end = farthest

        return jumps

jg = JumpGame()
nums = [2, 3, 1, 1, 4]
_nums = [3, 2, 1, 0, 4]
print(jg.canJump(nums))
print(jg.canJump(_nums))
print(jg.jump(nums))

